File: iosApp/iosApp/Presentation/Main/CafeManagement/MenuGoodsEdit/MenuGoodsEditViewModel.swift
//
//  MenuGoodsEditViewModel.swift
//  ConCafe
//
//  Created by 홍희표 on 2026/03/11.
//

import Foundation
import Combine
import Shared

@MainActor
final class MenuGoodsEditViewModel: ObservableObject {
    private let cafeId: String

    private let itemId: String?

    private let getCafeDetailUseCase: GetCafeDetailUseCase

    private let upsertCafeMenuGoodsUseCase: UpsertCafeMenuGoodsUseCase

    private let uploadImageUseCase: UploadImageUseCase

    @Published private(set) var uiState = MenuGoodsEditUiState()

    let event = PassthroughSubject<MenuGoodsEditEvent, Never>()

    private func loadInitialValue() {
        guard let itemId else {
            uiState.isLoading = false
            uiState.isEditMode = false
            return
        }
        uiState.isLoading = true
        uiState.infoMessageKey = nil

        Task {
            do {
                let result = try await getCafeDetailUseCase.invoke(cafeId: cafeId)

                if let success = result as? AppResultSuccess<AnyObject>,
                   let feed = success.data as? CafeDetailFeed {
                    let detail = feed.detail
                    uiState.detail = detail

                    if let menu = detail.menus.first(where: { $0.id == itemId }) {
                        applyMenu(menu)
                    } else if let goods = detail.goods.first(where: { $0.id == itemId }) {
                        applyGoods(goods)
                    } else {
                        showInfoAndStop(MessageKey.itemNotFound)
                    }
                } else {
                    showInfoAndStop(MessageKey.loadFailed)
                }
            } catch {
                if Task.isCancelled { return }
                showInfoAndStop(MessageKey.loadFailed)
            }
        }
    }

    private func saveItem() {
        guard !uiState.itemName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            uiState.infoMessageKey = MessageKey.enterName
            return
        }

        guard !uiState.price.isEmpty, let price = Int32(uiState.price) else {
            uiState.infoMessageKey = MessageKey.enterPrice
            return
        }

        uiState.isSaving = true
        uiState.infoMessageKey = nil

        Task {
            do {
                let uploadedImageUrl = try await uploadImageIfNeeded(uiState.imageUrl, folder: "cafe-items")
                let update = CafeMenuGoodsUpsert(
                    cafeId: cafeId,
                    itemId: itemId,
                    name: uiState.itemName,
                    price: price,
                    category: uiState.selectedCategoryId,
                    description: uiState.description,
                    isInStock: uiState.isInStock,
                    imageUrl: uploadedImageUrl
                )
                let result = try await upsertCafeMenuGoodsUseCase.invoke(update: update)

                if let success = result as? AppResultSuccess<AnyObject>,
                   let detail = success.data as? CafeDetail {
                    uiState.detail = detail
                    uiState.isSaving = false
                    uiState.isLoading = false
                    uiState.isEditMode = true
                    event.send(.navigateBack)
                } else {
                    uiState.isSaving = false
                    uiState.infoMessageKey = MessageKey.saveFailed
                }
            } catch {
                if Task.isCancelled { return }
                uiState.isSaving = false
                uiState.infoMessageKey = MessageKey.saveFailed
            }
        }
    }

    private func applyMenu(_ menu: CafeMenu) {
        uiState.isLoading = false
        uiState.isEditMode = true
        uiState.itemName = menu.name
        uiState.price = String(menu.price)
        uiState.selectedCategoryId = normalizeCategoryId(menu.category)
        uiState.description = menu.desc
        uiState.isInStock = menu.isAvailable
        uiState.imageUrl = menu.image
        uiState.infoMessageKey = nil
    }

    private func applyGoods(_ goods: Goods) {
        uiState.isLoading = false
        uiState.isEditMode = true
        uiState.itemName = goods.name
        uiState.price = String(goods.price)
        uiState.selectedCategoryId = "goods"
        uiState.description = "카페 굿즈 판매 항목"
        uiState.isInStock = goods.stock > 0
        uiState.imageUrl = goods.image
        uiState.infoMessageKey = nil
    }

    private func showInfoAndStop(_ messageKey: String) {
        uiState.isLoading = false
        uiState.infoMessageKey = messageKey
    }

    func onAction(_ action: MenuGoodsEditAction) {
        switch action {
        case .clickBack:
            event.send(.navigateBack)
        case .clickPhotoUpload:
            uiState.infoMessageKey = MessageKey.imageUploadNextStep
        case .selectPhoto(let imageUrl):
            uiState.imageUrl = imageUrl
        case .changeName(let value):
            uiState.itemName = value
        case .changePrice(let value):
            uiState.price = String(value.filter(\.isNumber))
        case .selectCategory(let category):
            uiState.selectedCategoryId = normalizeCategoryId(category)
        case .changeDescription(let value):
            uiState.description = value
        case .toggleStock(let isInStock):
            uiState.isInStock = isInStock
        case .clickSave:
            saveItem()
        case .dismissInfoMessage:
            uiState.infoMessageKey = nil
        }
    }

    init(
        cafeId: String,
        itemId: String? = nil,
        getCafeDetailUseCase: GetCafeDetailUseCase = KoinInitializerKt.resolveGetCafeDetailUseCase(),
        upsertCafeMenuGoodsUseCase: UpsertCafeMenuGoodsUseCase = KoinInitializerKt.resolveUpsertCafeMenuGoodsUseCase(),
        uploadImageUseCase: UploadImageUseCase = KoinInitializerKt.resolveUploadImageUseCase()
    ) {
        self.cafeId = cafeId
        self.itemId = itemId
        self.getCafeDetailUseCase = getCafeDetailUseCase
        self.upsertCafeMenuGoodsUseCase = upsertCafeMenuGoodsUseCase
        self.uploadImageUseCase = uploadImageUseCase

        loadInitialValue()
    }

    private func normalizeCategoryId(_ categoryId: String) -> String {
        let normalized = categoryId.lowercased()
        switch normalized {
        case "drink", "food", "dessert", "goods":
            return normalized
        default:
            return "drink"
        }
    }

    private func uploadImageIfNeeded(_ imageUrl: String?, folder: String) async throws -> String? {
        guard let imageUrl, !imageUrl.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return nil
        }
        let result = try await uploadImageUseCase.invoke(localPath: imageUrl, folder: folder)
        if let success = result as? AppResultSuccess<AnyObject>, let data = success.data as? String {
            return data
        }
        if let failure = result as? AppResultFailure, let validation = failure.error as? AppErrorValidationFailed {
            throw NSError(domain: "MenuGoodsEdit", code: 1, userInfo: [NSLocalizedDescriptionKey: validation.reason])
        }
        throw NSError(domain: "MenuGoodsEdit", code: 1, userInfo: [NSLocalizedDescriptionKey: String(localized: String.LocalizationValue(MessageKey.imageUploadFailed), table: "Localizable")])
    }

    private enum MessageKey {
        static let itemNotFound = "menugoods_edit_info_item_not_found"
        static let loadFailed = "menugoods_edit_info_load_failed"
        static let enterName = "menugoods_edit_info_enter_name"
        static let enterPrice = "menugoods_edit_info_enter_price"
        static let saveFailed = "menugoods_edit_info_save_failed"
        static let imageUploadNextStep = "menugoods_edit_info_image_upload_next_step"
        static let imageUploadFailed = "menugoods_edit_info_image_upload_failed"
    }
}

File: composeApp/src/commonMain/kotlin/com/hhp227/concafe/presentation/main/cafemanagement/menugoodsedit/MenuGoodsEditViewModel.kt
package com.hhp227.concafe.presentation.main.cafemanagement.menugoodsedit

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import com.hhp227.concafe.domain.common.AppResult
import com.hhp227.concafe.domain.model.CafeMenuGoodsUpsert
import com.hhp227.concafe.domain.model.CafeMenu
import com.hhp227.concafe.domain.model.Goods
import com.hhp227.concafe.domain.usecase.GetCafeDetailUseCase
import com.hhp227.concafe.domain.usecase.UploadImageUseCase
import com.hhp227.concafe.domain.usecase.UpsertCafeMenuGoodsUseCase

class MenuGoodsEditViewModel(
    private val cafeId: String,
    private val itemId: String?,
    private val getCafeDetailUseCase: GetCafeDetailUseCase,
    private val upsertCafeMenuGoodsUseCase: UpsertCafeMenuGoodsUseCase,
    private val uploadImageUseCase: UploadImageUseCase
) : ViewModel() {
    private val _uiState = MutableStateFlow(MenuGoodsEditUiState())
    val uiState = _uiState.asStateFlow()

    private val _event = MutableSharedFlow<MenuGoodsEditEvent>(replay = 0)
    val event = _event.asSharedFlow()

    private fun loadInitialValue() {
        if (itemId == null) {
            _uiState.update {
                it.copy(
                    isLoading = false,
                    isEditMode = false
                )
            }
            return
        }
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, infoMessageKey = null) }
            when (val result = getCafeDetailUseCase.invoke(cafeId)) {
                is AppResult.Success -> {
                    val detail = result.data.detail
                    val menu = detail.menus.firstOrNull { it.id == itemId }
                    val goods = detail.goods.firstOrNull { it.id == itemId }

                    when {
                        menu != null -> applyMenu(menu)
                        goods != null -> applyGoods(goods)
                        else -> showInfoAndStop(MSG_ITEM_NOT_FOUND)
                    }
                }
                is AppResult.Failure -> showInfoAndStop(MSG_LOAD_FAILED)
            }
        }
    }

    private fun clickBack() {
        viewModelScope.launch {
            _event.emit(MenuGoodsEditEvent.NavigateBack)
        }
    }

    private fun clickSave() {
        val currentState = _uiState.value
        when {
            currentState.itemName.isBlank() -> showInfo(MSG_ENTER_NAME)
            currentState.price.isBlank() -> showInfo(MSG_ENTER_PRICE)
            currentState.price.toIntOrNull() == null -> showInfo(MSG_PRICE_NUMBER_ONLY)
            else -> {
                _uiState.update { it.copy(infoMessageKey = null) }
                viewModelScope.launch {
                    val uploadedImageUrl = uploadImage(currentState.imageUrl, "cafe-items")
                    if (!currentState.imageUrl.isNullOrBlank() && uploadedImageUrl == null) {
                        return@launch
                    }
                    when (
                        val result = upsertCafeMenuGoodsUseCase.invoke(
                            CafeMenuGoodsUpsert(
                                cafeId = cafeId,
                                itemId = itemId,
                                name = currentState.itemName.trim(),
                                price = currentState.price.toInt(),
                                category = currentState.selectedCategoryId,
                                description = currentState.description.trim(),
                                isInStock = currentState.isInStock,
                                imageUrl = uploadedImageUrl
                            )
                        )
                    ) {
                        is AppResult.Success -> {
                            _event.emit(MenuGoodsEditEvent.NavigateBack)
                        }
                        is AppResult.Failure -> showInfo(MSG_SAVE_FAILED)
                    }
                }
            }
        }
    }

    private fun showInfo(messageKey: String) {
        _uiState.update { it.copy(infoMessageKey = messageKey) }
    }

    private suspend fun uploadImage(imageUrl: String?, folder: String): String? {
        if (imageUrl.isNullOrBlank()) return null
        return when (val result = uploadImageUseCase.invoke(imageUrl, folder)) {
            is AppResult.Success -> result.data
            is AppResult.Failure -> {
                _uiState.update {
                    it.copy(
                        infoMessageKey = MSG_IMAGE_UPLOAD_FAILED
                    )
                }
                null
            }
        }
    }

    private fun applyMenu(menu: CafeMenu) {
        _uiState.update {
            it.copy(
                isLoading = false,
                isEditMode = true,
                itemName = menu.name,
                price = menu.price.toString(),
                selectedCategoryId = when (menu.category.lowercase()) {
                    "food", "dessert", "goods", "drink" -> menu.category.lowercase()
                    else -> "drink"
                },
                description = menu.desc,
                isInStock = menu.isAvailable,
                imageUrl = menu.image
            )
        }
    }

    private fun applyGoods(goods: Goods) {
        _uiState.update {
            it.copy(
                isLoading = false,
                isEditMode = true,
                itemName = goods.name,
                price = goods.price.toString(),
                selectedCategoryId = "goods",
                description = "카페 굿즈 판매 항목",
                isInStock = goods.stock > 0,
                imageUrl = goods.image
            )
        }
    }

    private fun showInfoAndStop(messageKey: String) {
        _uiState.update {
            it.copy(
                isLoading = false,
                infoMessageKey = messageKey
            )
        }
    }

    fun onAction(action: MenuGoodsEditAction) {
        when (action) {
            MenuGoodsEditAction.ClickBack -> clickBack()
            MenuGoodsEditAction.ClickPhotoUpload -> showInfo(MSG_IMAGE_UPLOAD_NEXT_STEP)
            is MenuGoodsEditAction.ChangeName -> _uiState.update { it.copy(itemName = action.value) }
            is MenuGoodsEditAction.ChangePrice -> _uiState.update { it.copy(price = action.value.filter(Char::isDigit)) }
            is MenuGoodsEditAction.SelectPhoto -> _uiState.update { it.copy(imageUrl = action.imageUrl) }
            is MenuGoodsEditAction.SelectCategory -> _uiState.update { it.copy(selectedCategoryId = action.categoryId) }
            is MenuGoodsEditAction.ChangeDescription -> _uiState.update { it.copy(description = action.value) }
            is MenuGoodsEditAction.ToggleStock -> _uiState.update { it.copy(isInStock = action.isInStock) }
            MenuGoodsEditAction.ClickSave -> clickSave()
            MenuGoodsEditAction.DismissInfoMessage -> _uiState.update { it.copy(infoMessageKey = null) }
        }
    }

    init {
        loadInitialValue()
    }

    companion object {
        private const val MSG_ITEM_NOT_FOUND = "menugoods_edit_info_item_not_found"
        private const val MSG_LOAD_FAILED = "menugoods_edit_info_load_failed"
        private const val MSG_ENTER_NAME = "menugoods_edit_info_enter_name"
        private const val MSG_ENTER_PRICE = "menugoods_edit_info_enter_price"
        private const val MSG_PRICE_NUMBER_ONLY = "menugoods_edit_info_price_number_only"
        private const val MSG_SAVE_FAILED = "menugoods_edit_info_save_failed"
        private const val MSG_IMAGE_UPLOAD_FAILED = "menugoods_edit_info_image_upload_failed"
        private const val MSG_IMAGE_UPLOAD_NEXT_STEP = "menugoods_edit_info_image_upload_next_step"
    }
}
